File: src/examples/org/apache/hama/examples/MatrixMultiplication.java
/**
 * Copyright 2007 The Apache Software Foundation
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.hama.examples;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import org.apache.hadoop.hbase.HColumnDescriptor;
import org.apache.hadoop.hbase.HTableDescriptor;
import org.apache.hadoop.hbase.client.HBaseAdmin;
import org.apache.hadoop.hbase.client.Scan;
import org.apache.hadoop.hbase.mapreduce.TableMapReduceUtil;
import org.apache.hadoop.hbase.util.Bytes;
import org.apache.hadoop.io.BytesWritable;
import org.apache.hadoop.io.IntWritable;
import org.apache.hadoop.io.MapWritable;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hama.Constants;
import org.apache.hama.HamaAdmin;
import org.apache.hama.HamaAdminImpl;
import org.apache.hama.examples.mapreduce.*;
import org.apache.hama.io.BlockID;
import org.apache.hama.matrix.DenseMatrix;
import org.apache.hama.matrix.Matrix;
import org.apache.hama.matrix.SparseMatrix;
import org.apache.hama.util.RandomVariable;

public class MatrixMultiplication extends AbstractExample {
  public static void main(String[] args) throws IOException {
    if (args.length < 2) {
      System.out
          .println("mult [-m maps] [-r reduces] <matrix A> <matrix B> [blocks]");
      System.exit(-1);
    } else {
      parseArgs(args);
    }

    String matrixA = ARGS.get(0);
    String matrixB = ARGS.get(1);

    HamaAdmin admin = new HamaAdminImpl(conf);
    Matrix a = admin.getMatrix(matrixA);
    Matrix b = admin.getMatrix(matrixB);

    if (!a.getType().equals(b.getType())) {
      System.out.println(a.getType() + " != " + b.getType());
      System.exit(-1);
    }

    Matrix c;
    if (a.getType().equals("SparseMatrix")) {
      if (ARGS.size() > 2) {
        System.out
            .println("NOTE: You can't use the block algorithm for sparse matrix multiplication.");
      }
      c = ((SparseMatrix) a).mult(b);
    } else {
      if (ARGS.size() > 2) {
        c = mult(a, b, Integer.parseInt(ARGS.get(2)));
      } else {
        c = mult(a, b);
      }
    }

    for (int i = 0; i < 2; i++) {
      System.out.println("c(" + 0 + ", " + i + ") : " + c.get(0, i));
    }
    System.out.println("...");
  }

  /**
   * C = A*B using iterative method
   * 
   * @param B
   * @return C
   * @throws IOException
   */
  public static DenseMatrix mult(Matrix A, Matrix B) throws IOException {
    ensureForMultiplication(A, B);
    int columns = 0;
    if (B.getColumns() == 1 || A.getColumns() == 1)
      columns = 1;
    else
      columns = A.getColumns();

    DenseMatrix result = new DenseMatrix(conf, A.getRows(), columns);
    List<Job> jobId = new ArrayList<Job>();

    for (int i = 0; i < A.getRows(); i++) {
      Job job = new Job(conf, "multiplication MR job : " + result.getPath()
          + " " + i);

      Scan scan = new Scan();
      scan.addFamily(Constants.COLUMNFAMILY);
      job.getConfiguration()
          .set(DenseMatrixVectorMultMap.MATRIX_A, A.getPath());
      job.getConfiguration().setInt(DenseMatrixVectorMultMap.ITH_ROW, i);

      TableMapReduceUtil.initTableMapperJob(B.getPath(), scan,
          DenseMatrixVectorMultMap.class, IntWritable.class, MapWritable.class,
          job);
      TableMapReduceUtil.initTableReducerJob(result.getPath(),
          DenseMatrixVectorMultReduce.class, job);
      try {
        job.waitForCompletion(false);
        jobId.add(job);
      } catch (InterruptedException e) {
        e.printStackTrace();
      } catch (ClassNotFoundException e) {
        e.printStackTrace();
      }
    }

    while (checkAllJobs(jobId) == false) {
      try {
        Thread.sleep(1000);
      } catch (InterruptedException e) {
        // TODO Auto-generated catch block
        e.printStackTrace();
      }
    }

    return result;
  }

  /**
   * C = A * B using Blocking algorithm
   * 
   * @param B
   * @param blocks the number of blocks
   * @return C
   * @throws IOException
   */
  public static DenseMatrix mult(Matrix A, Matrix B, int blocks)
      throws IOException {
    ensureForMultiplication(A, B);
    
    String collectionTable = "collect_" + RandomVariable.randMatrixPath();
    HTableDescriptor desc = new HTableDescriptor(collectionTable);
    desc.addFamily(new HColumnDescriptor(Bytes.toBytes(Constants.BLOCK)));
    new HBaseAdmin(conf).createTable(desc);

    collectBlocksMapRed(A, collectionTable, blocks, true);
    collectBlocksMapRed(B, collectionTable, blocks, false);

    DenseMatrix result = new DenseMatrix(conf, A.getRows(), A.getColumns());

    Job job = new Job(conf, "multiplication MR job : " + result.getPath());

    Scan scan = new Scan();
    scan.addFamily(Bytes.toBytes(Constants.BLOCK));

    TableMapReduceUtil.initTableMapperJob(collectionTable, scan,
        BlockMultMap.class, BlockID.class, BytesWritable.class, job);
    TableMapReduceUtil.initTableReducerJob(result.getPath(),
        BlockMultReduce.class, job);

    try {
      job.waitForCompletion(true);
    } catch (InterruptedException e) {
      e.printStackTrace();
    } catch (ClassNotFoundException e) {
      e.printStackTrace();
    }

    new HamaAdminImpl(conf, new HBaseAdmin(conf)).delete(collectionTable);
    return result;
  }

  /**
   * Collect Blocks
   * 
   * @param path a input path
   * @param collectionTable the collection table
   * @param blockNum the number of blocks
   * @param bool
   * @throws IOException
   */
  public static void collectBlocksMapRed(Matrix m, String collectionTable,
      int blockNum, boolean bool) throws IOException {
    double blocks = Math.pow(blockNum, 0.5);
    if (!String.valueOf(blocks).endsWith(".0"))
      throw new IOException("can't divide.");

    int block_size = (int) blocks;
    Job job = new Job(conf, "Blocking MR job" + m.getPath());

    Scan scan = new Scan();
    scan.addFamily(Constants.COLUMNFAMILY);

    job.getConfiguration().set(CollectBlocksMapper.BLOCK_SIZE,
        String.valueOf(block_size));
    job.getConfiguration().set(CollectBlocksMapper.ROWS,
        String.valueOf(m.getRows()));
    job.getConfiguration().set(CollectBlocksMapper.COLUMNS,
        String.valueOf(m.getColumns()));
    job.getConfiguration().setBoolean(CollectBlocksMapper.MATRIX_POS, bool);

    TableMapReduceUtil.initTableMapperJob(m.getPath(), scan,
        org.apache.hama.examples.mapreduce.CollectBlocksMapper.class, BlockID.class,
        MapWritable.class, job);
    TableMapReduceUtil.initTableReducerJob(collectionTable,
        org.apache.hama.examples.mapreduce.CollectBlocksReducer.class, job);

    try {
      job.waitForCompletion(true);
    } catch (InterruptedException e) {
      e.printStackTrace();
    } catch (ClassNotFoundException e) {
      e.printStackTrace();
    }
  }

  public static void ensureForMultiplication(Matrix A, Matrix m)
      throws IOException {
    if (A.getColumns() != m.getRows()) {
      throw new IOException("A's columns should equal with B's rows while A*B.");
    }
  }

  public static boolean checkAllJobs(List<Job> jobId) throws IOException {
    Iterator<Job> it = jobId.iterator();
    boolean allTrue = true;
    while (it.hasNext()) {
      if (!it.next().isComplete()) {
        allTrue = false;
      }
    }

    return allTrue;
  }
}

File: c++/src/main/native/examples/impl/matrixmultiplication.cc
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "hama/Pipes.hh"
#include "hama/TemplateFactory.hh"
#include "hadoop/StringUtils.hh"
#include "DenseDoubleVector.hh"

#include <time.h>
#include <math.h>
#include <string>
#include <string>
#include <iostream>
#include <sstream>

using std::string;
using std::cout;

using HamaPipes::BSP;
using HamaPipes::BSPJob;
using HamaPipes::Partitioner;
using HamaPipes::BSPContext;
using namespace HadoopUtils;

using math::DenseDoubleVector;

class MatrixMultiplicationBSP: public BSP {
private:
  string masterTask;
  int seqFileID;
  string HAMA_MAT_MULT_B_PATH;
public:
  MatrixMultiplicationBSP(BSPContext& context) {
    seqFileID = 0;
    HAMA_MAT_MULT_B_PATH = "hama.mat.mult.B.path";
  }
  
  void setup(BSPContext& context) {
    // Choose one as a master
    masterTask = context.getPeerName(context.getNumPeers() / 2);
    
    reopenMatrixB(context);
  }
  
  void bsp(BSPContext& context) {
    
    string aRowKey;
    string aRowVectorStr;
    // while for each row of matrix A
    while(context.readNext(aRowKey, aRowVectorStr)) {
      
      DenseDoubleVector *aRowVector = new DenseDoubleVector(aRowVectorStr);
      DenseDoubleVector *colValues = NULL;
      
      string bColKey;
      string bColVectorStr;
      
      // while for each col of matrix B
      while (context.sequenceFileReadNext(seqFileID,bColKey,bColVectorStr)) {
        
        DenseDoubleVector *bColVector = new DenseDoubleVector(bColVectorStr);
        
        if (colValues == NULL)
          colValues = new DenseDoubleVector(bColVector->getDimension());
        
        double dot = aRowVector->dot(bColVector);
        
        colValues->set(toInt(bColKey), dot);
      }
      
      // Submit one calculated row
      std::stringstream message;
      message << aRowKey << ":" << colValues->toString();
      context.sendMessage(masterTask, message.str());
      
      reopenMatrixB(context);
    }
    
    context.sequenceFileClose(seqFileID);
    context.sync();
  }
  
  void cleanup(BSPContext& context) {
    if (context.getPeerName().compare(masterTask)==0) {
      
      int msgCount = context.getNumCurrentMessages();
      
      for (int i=0; i<msgCount; i++) {
        
        string received = context.getCurrentMessage();
        //key:value1,value2,value3
        int pos = (int)received.find(":");
        string key = received.substr(0,pos);
        string values = received.substr(pos+1,received.length());
        
        context.write(key, values);
      }
    }
  }
  
  void reopenMatrixB(BSPContext& context) {
    if (seqFileID!=0) {
      context.sequenceFileClose(seqFileID);
    }
    
    const BSPJob* job = context.getBSPJob();
    string path = job->get(HAMA_MAT_MULT_B_PATH);
    
    seqFileID = context.sequenceFileOpen(path,"r",
                                         "org.apache.hadoop.io.IntWritable",
                                         "org.apache.hama.ml.writable.VectorWritable");
  }
  
};

class MatrixRowPartitioner: public Partitioner {
public:
  MatrixRowPartitioner(BSPContext& context) { }
  
  int partition(const string& key,const string& value, int32_t numTasks) {
    return toInt(key) % numTasks;
  }
};

int main(int argc, char *argv[]) {
  return HamaPipes::runTask(HamaPipes::TemplateFactory<MatrixMultiplicationBSP,MatrixRowPartitioner>());
}
